### PR TITLE
Reject the batch with incorrect magic attribute with the valid response

### PR DESCRIPTION
**Pull request by clock21am**

### CODE REVIEW

File: batch/batch.go [hunk 1]
```diff
@@ -119,15 +119,18 @@ func (b *Builder) Build(now time.Time) (*Batch, error) {
 
 var (
 	CorruptedBatchError = errors.New("batch crc does not match bytes")
+	UnsupportedProtocolError = errors.New(" protocol version with magic value 2 is only supported")
```
Comment (twalwyn) [inline batch/batch.go:122]: " protocol version with magic value 2 is only supported" -> "batch magic value is not 2"

File: batch/batch.go [hunk 2]
```diff
@@ -119,15 +119,18 @@ func (b *Builder) Build(now time.Time) (*Batch, error) {
...
 	crc32c              = crc32.MakeTable(crc32.Castagnoli)
 )
-
 // Unmarshal the batch. On error batch is nil. If there is an error, it is most
 // likely because the crc failed. In that case there is no way to tell how many
 // records there were in the batch (and to adjust offsets accordingly).
 func Unmarshal(b []byte) (*Batch, error) {
 	buf := bytes.NewBuffer(b)
 	batch := &Batch{}
+	if b[unsafe.Offsetof(batch.Magic)] != 2 {
```
Comment (twalwyn) [inline batch/batch.go:131]: i don't really understand why we need to retrieve the offset of the field from the struct using `unsafe`. should we keep it simple and have a constant containing the index of the magic byte?

**Review by twalwyn — CHANGES_REQUESTED** [review]
thanks for making the PR!

File: batch/batch.go (hunk 2 above)
Comment (clock21am) [reply batch/batch.go:131]: I feel that too

File: batch/batch.go [hunk 3]
```diff
@@ -118,14 +118,30 @@ func (b *Builder) Build(now time.Time) (*Batch, error) {
 }
 
 var (
-	CorruptedBatchError = errors.New("batch crc does not match bytes")
-	crc32c              = crc32.MakeTable(crc32.Castagnoli)
+	CorruptedBatchError      = errors.New("batch crc does not match bytes")
+	UnsupportedProtocolError = errors.New("magic value is not 2")
```
Comment (twalwyn) [inline batch/batch.go:122]:
```suggestion
	UnsupportedMagicError = errors.New("magic value is not 2")
```

**Review by twalwyn — APPROVED** [review]

**Comment by mkocikowski** [comment]
thanks for doing this. please add a comment to verifyMagicByte explaining that prior to version 0.11 kafka used message sets (https://kafka.apache.org/documentation/#messageset) and starting with 0.11 it uses record batches (https://kafka.apache.org/documentation/#recordbatch). libkafka only supports record batches. but, apparently if a pre-0.11 client writes to kafka, the old message set format is used, and that is what is returned by fetch calls - but libkafka doesn't support that. or something along these lines.

**Review by mkocikowski — APPROVED** [review]
see my other comment on explaining more about "why" (message set vs record batch) but looks good to me

**Review by mkocikowski — APPROVED** [review]

**Pull request merged**